packages/darklang/stdlib/json.dark

```diff
@@ -0,0 +1,113 @@
+module Darklang =
+  module Stdlib =
+    module Json =
+      module ParseError =
+        module JsonPath =
+          module Part =
+            type Part =
+              | Root
+              | Index of Int
+              | Field of String
+
+            let toString (part: Part) : String =
+              match part with
+              | Root -> "root"
+              | Field(name) -> "." ++ name
+              | Index(index) -> "[" ++ Int.toString index ++ "]"
+
+          type JsonPath = List<Part.Part>
+
+          let toString (path: JsonPath) : String =
+            path |> List.reverse |> List.map Part.toString |> String.join ""
+
+
+        type ParseError =
+          /// The json string can't be parsed as the given type.
+          | CantMatchWithType of
+            PACKAGE.Darklang.LanguageTools.RuntimeTypes.TypeReference *
+            String *
+            JsonPath.JsonPath
+          | EnumExtraField of rawJson: String * JsonPath.JsonPath
+          | EnumMissingField of
+            PACKAGE.Darklang.LanguageTools.RuntimeTypes.TypeReference *
+            index: Int *
+            JsonPath.JsonPath
+          | EnumInvalidCasename of
+            PACKAGE.Darklang.LanguageTools.RuntimeTypes.TypeReference *
+            caseName: String *
+            JsonPath.JsonPath
+          | EnumTooManyCases of
+            PACKAGE.Darklang.LanguageTools.RuntimeTypes.TypeReference *
+            cases: List<String> *
+            JsonPath.JsonPath
+          | RecordDuplicateField of fieldName: String * JsonPath.JsonPath
+          | RecordMissingField of fieldName: String * JsonPath.JsonPath
+          | NotJson
```

Comment:
> One open question I have is whether putting a TypeReference in the type is a good idea. This isn't a RuntimeError that will be handled by the runtime, it's a user facing error. So I'm not quite sure what to do with this.

Hmm. Is the concern here that users would then be exposed to `TypeReference`, which feels odd?

Reply:
yes, exactly.